packages/shared/src/features/onboarding/hooks/useFunnelTracking.ts

```diff
@@ -0,0 +1,172 @@
+import { useAtomValue } from 'jotai';
+import { useLogContext } from '../../../contexts/LogContext';
+import type {
+  FunnelStep,
+  FunnelJSON,
+  FunnelStepTransitionType,
+} from '../types/funnel';
+import type { FunnelEvent } from '../types/funnelEvents';
+import { FunnelEventName } from '../types/funnelEvents';
+import {
+  getFunnelStepByPosition,
+  funnelPositionAtom,
+} from '../store/funnelStore';
+
+type TrackOnMouseCapture = MouseEventHandler<HTMLElement>;
+type TrackOnScroll = () => void;
+type TrackOnEvent = (event: FunnelEvent) => void;
+export type TrackOnNavigate = (event: {
+  from: FunnelStep['id'];
+  to: FunnelStep['id'];
+  timeDuration: number;
+  type: FunnelStepTransitionType;
+}) => void;
+
+interface USeFunnelTrackingProps {
+  funnel: FunnelJSON;
+}
+
+interface UseFunnelTrackingReturn {
+  trackOnClickCapture: TrackOnMouseCapture;
+  trackOnHoverCapture: TrackOnMouseCapture;
+  trackFunnelEvent: TrackOnEvent;
+  trackOnNavigate: TrackOnNavigate;
+  trackOnScroll: TrackOnScroll;
+}
+
+const trackOnMouseCapture = ({
+  selector,
+  eventName,
+  trackFunnelEvent,
+}: {
+  selector: string;
+  eventName:
+    | FunnelEventName.ClickFunnelElement
+    | FunnelEventName.HoverFunnelElement;
+  trackFunnelEvent: TrackOnEvent;
+}): TrackOnMouseCapture => {
+  return (event) => {
+    if (!(event.target instanceof HTMLElement)) {
+      return;
+    }
+
+    const trackedElement = event.target.closest(selector);
+
+    if (!trackedElement) {
+      return;
+    }
+
+    trackFunnelEvent({
+      name: eventName,
+      details: {
+        target_type:
+          trackedElement.tagName.toLowerCase() as keyof HTMLElementTagNameMap,
+        ...(trackedElement.id ? { target_id: trackedElement.id } : {}),
+      },
+    });
+  };
+};
+
+export const useFunnelTracking = ({
+  funnel,
+}: USeFunnelTrackingProps): UseFunnelTrackingReturn => {
+  const { logEvent } = useLogContext();
+  const position = useAtomValue(funnelPositionAtom);
+  const step = useMemo(
+    () => getFunnelStepByPosition(funnel, position),
+    [funnel, position],
+  );
+
+  const trackFunnelEvent: TrackOnEvent = useCallback(
+    (event) => {
+      const commonTrackingProps = {
+        funnel_id: funnel.id,
+        funnel_version: funnel.version,
+        session_id: '', // todo: implement sessionId
+        step_id: step?.id,
+        step_type: step?.type,
+      };
+
+      logEvent({
+        event_name: event.name,
+        extra: JSON.stringify({
+          ...commonTrackingProps,
+          ...('details' in event ? event.details : {}),
+        }),
+      });
+    },
+    [funnel, logEvent, step],
+  );
+
+  const trackOnClickCapture: TrackOnMouseCapture = trackOnMouseCapture({
+    selector: '[data-track-click]',
+    eventName: FunnelEventName.ClickFunnelElement,
+    trackFunnelEvent,
+  });
+
+  const trackOnHoverCapture: TrackOnMouseCapture = trackOnMouseCapture({
+    selector: '[data-track-hover]',
```

Comment:
i think every element that should be tracked for hover will be tracked for click. should we track the same selector for both?

Reply:
Done ✔️